timesketch/lib/datastores/opensearch.py

```diff
@@ -1271,3 +1296,505 @@ def flush_queued_events(self, retry_count=0):
+
+        Raises:
+            ValueError: If indices_for_pit is empty or num_slices is invalid.
+            errors.DataStoreQueryError: For critical issues during the export.
+        """
+        if not indices_for_pit:
+            raise ValueError("indices_for_pit cannot be empty.")
+
+        # Getting default values if not provided otherwise by the caller.
+        effective_sort_criteria = sort_criteria or _DEFAULT_PIT_SORT_CRITERIA
+        effective_page_size = (
+            page_size if page_size is not None else self.sliced_export_default_page_size
+        )
+        effective_pit_keep_alive = (
+            pit_keep_alive
+            if pit_keep_alive is not None
+            else self.sliced_export_pit_keep_alive
+        )
+        effective_num_slices = (
+            num_slices
+            if num_slices is not None
+            else self.sliced_export_num_slices_default
+        )
+        effective_worker_timeout = (
+            request_timeout_per_slice
+            if request_timeout_per_slice is not None
+            else self.sliced_export_request_timeout_default
+        )
+
+        if not 1 <= effective_num_slices <= 1024:  # Validate after resolving default
+            raise ValueError("num_slices must be between 1 and 1024.")
+
+        # Use self.sliced_export_queue_buffer_factor for queue size calculation
+        queue_max_items = (
+            effective_page_size
+            * effective_num_slices
+            * self.sliced_export_queue_buffer_factor
+        )
+        results_queue: queue.Queue = queue.Queue(maxsize=queue_max_items)
+
+        threads: List[threading.Thread] = []
+        stop_event = threading.Event()
+
+        os_logger.info(
+            "Starting sliced export from indices: [%s] with %d slices, page_size: %d.",
+            ", ".join(indices_for_pit),
+            effective_num_slices,
+            effective_page_size,
+        )
+
+        active_workers = effective_num_slices
+
+        try:
+            for i in range(effective_num_slices):
+                thread = threading.Thread(
+                    target=self._export_slice_worker,
+                    args=(
+                        i,
+                        effective_num_slices,
+                        indices_for_pit,
+                        copy.deepcopy(base_query_body),
+                        effective_sort_criteria,
+                        effective_page_size,
+                        effective_pit_keep_alive,
+                        results_queue,
+                        stop_event,
+                        effective_worker_timeout,
+                    ),
+                    # Allows main thread to exit even if workers hang
+                    # (though join is preferred)
+                    daemon=True,
+                )
+                threads.append(thread)
+                thread.start()
+
+            while active_workers > 0:
+                if stop_event.is_set() and results_queue.empty():
```

Comment:
do you also want to stop the while loop when `active_workers <= 0`

Reply:
Yes, if the `active_workers` get negative at any time (don't know how) the loop should also stop.